File: client_go_gorilla_websocket/attacker.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"net/http"
	"os"
	"time"

	"github.com/dustin/go-humanize"
	"github.com/gorilla/websocket"
	"gopkg.in/vmihailenco/msgpack.v2"
)

type Config struct {
	URL      string
	Loops    int
	Interval int // msec
	Keep     int // sec
}

type TAT struct {
	Target    string // client address "ip:port"
	MsgNum    int
	isStart   bool
	StartTime time.Time
	Sum       float64
	Min       float64
	Max       float64
}

type StressTestInfo struct {
	StartTime       time.Time
	AllSendByteSize int
	AllRecvByteSize int
	ConnExecTimeSum float64 // for avarage time
	ConnExecTimeMin float64
	ConnExecTimeMax float64
	ConnectionNum   int
	MessageTAT      TAT
	Config          *Config
}

func (st *StressTestInfo) send(ws *websocket.Conn, msg []map[string]interface{}) (err error) {
	packedMsg, err := msgpack.Marshal(msg)
	if err != nil {
		log.Printf("msgpack marshal error: %v", err)
		return err
	}
	if err = ws.WriteMessage(websocket.BinaryMessage, packedMsg); err != nil {
		log.Printf("write message: %v", err)
		return err
	}

	if st.MessageTAT.Target == ws.LocalAddr().String() {
		st.MessageTAT.isStart = true
		st.MessageTAT.MsgNum += 1
		st.MessageTAT.StartTime = time.Now()
	}

	st.AllSendByteSize += len(packedMsg)
	return err
}

func (st *StressTestInfo) execScenarioTest() (err error) {
	var wsDialer = &websocket.Dialer{}
	var wsHeader = http.Header{"Origin": {st.Config.URL}}

	for i := 0; i < st.Config.Loops; i++ {
		url := st.Config.URL + "ws"

		startTime := time.Now()
		ws, _, err := wsDialer.Dial(url, wsHeader)
		if err != nil {
			log.Printf("new client: %v", err)
			break
		}
		// for TAT
		if i == 0 {
			st.MessageTAT.Target = ws.LocalAddr().String()
		}

		endTime := time.Now()
		diffTime := endTime.Sub(startTime).Seconds()
		st.ConnExecTimeSum += diffTime
		if st.ConnExecTimeMin > diffTime {
			st.ConnExecTimeMin = diffTime
		}
		if st.ConnExecTimeMax < diffTime {
			st.ConnExecTimeMax = diffTime
		}

		// count up ws connection
		st.ConnectionNum += 1

		go st.RecvWithPingPong(ws)
		go st.execScenario(ws)

		time.Sleep(time.Duration(st.Config.Interval) * time.Millisecond)
	}

	fmt.Printf("keep %dsec\n", st.Config.Keep)
	time.Sleep(time.Duration(st.Config.Keep) * time.Second)

	fmt.Println(fmt.Sprintf("test end. max connection is %d", st.ConnectionNum))

	return nil
}

var MsgpackSendData = map[string]interface{}{
	"a": 1,
}

func (st *StressTestInfo) execScenario(ws *websocket.Conn) {
	// scenario
	for {
		_data := MsgpackSendData
		var msg []map[string]interface{}
		msg = append(msg, _data)
		st.send(ws, msg)

		time.Sleep(time.Duration(rand.Intn(7)+3) * time.Second)
	}
}

func (st *StressTestInfo) RecvWithPingPong(ws *websocket.Conn) (err error) {
	for {
		msgType, r, err := ws.NextReader()
		if err != nil {
			ws.Close()
			break
		}

		if msgType != websocket.BinaryMessage {
			continue
		}

		var buf []byte
		buf, err = ioutil.ReadAll(r)
		if err != nil {
			ws.Close()
			break
		}
		st.AllRecvByteSize += len(buf)

		if st.MessageTAT.isStart && st.MessageTAT.Target == ws.LocalAddr().String() {
			st.MessageTAT.isStart = false
			diffTime := time.Now().Sub(st.MessageTAT.StartTime).Seconds()
			st.MessageTAT.Sum += diffTime
			if st.MessageTAT.Min > diffTime {
				st.MessageTAT.Min = diffTime
			}
			if st.MessageTAT.Max < diffTime {
				st.MessageTAT.Max = diffTime
			}
		}
	}

	return nil
}

func (st *StressTestInfo) dumpInfo() {
	pid := os.Getpid()
	for {
		time.Sleep(time.Duration(10) * time.Second)
		now := time.Now()
		fmt.Printf("======= %v (elapsed: %v)\n", now, now.Sub(st.StartTime))
		fmt.Printf("tool's pid       : %d\n", pid)
		fmt.Printf("Target URL       : %s\n", st.Config.URL)
		fmt.Printf("Send Byte Size   : %s [byte] (%s)\n",
			humanize.Comma(int64(st.AllSendByteSize)), humanize.Bytes(uint64(st.AllSendByteSize)))
		fmt.Printf("Recive Byte Size : %s [byte] (%s)\n",
			humanize.Comma(int64(st.AllRecvByteSize)), humanize.Bytes(uint64(st.AllRecvByteSize)))
		fmt.Printf("Connection       : %d [conn]\n", st.ConnectionNum)
		fmt.Printf("Connect Time(avg): %.1f [ms]\n", st.ConnExecTimeSum/float64(st.ConnectionNum)*1000.)
		fmt.Printf("Connect Time(min): %.1f [ms]\n", st.ConnExecTimeMin*1000.)
		fmt.Printf("Connect Time(max): %.1f [ms]\n", st.ConnExecTimeMax*1000.)
		fmt.Printf("Message TAT (avg): %.1f [ms]\n", st.MessageTAT.Sum/float64(st.MessageTAT.MsgNum)*1000.)
		fmt.Printf("Message TAT (min): %.1f [ms]\n", st.MessageTAT.Min*1000.)
		fmt.Printf("Message TAT (max): %.1f [ms]\n", st.MessageTAT.Max*1000.)
	}
}

func main() {
	var conf string
	flag.StringVar(&conf, "config", "", "config file path")

	flag.Parse()

	// for stress test
	st := new(StressTestInfo)
	config := new(Config)
	configFile, err := ioutil.ReadFile(conf)
	if err != nil {
		log.Println("open config file: ", err)
		return
	}
	if err = json.Unmarshal(configFile, config); err != nil {
		log.Println("parse config file: ", err)
		return
	}
	fmt.Println(config)

	st.Config = config
	st.ConnExecTimeMin = 999999.9
	st.MessageTAT.Min = 999999.9
	st.StartTime = time.Now()
	go st.dumpInfo()

	st.execScenarioTest()
}

File: client_python_tornado/attacker.py
import argparse
import json
import os
import random
import threading
import time
from datetime import datetime
import humanize
import msgpack
from tornado import gen
from tornado.ioloop import PeriodicCallback, IOLoop
from tornado.websocket import websocket_connect

Sem = threading.Semaphore()
rtt_info = {
    'TargetID': None,
    'MsgNum': 0,
    'Sum': 0,
    'Min': 9999999999,
    'Max': 0,
}
testinfo = {
    'TargetURL': '',
    'StartTime': datetime.now(),
    'AllSendByteSize': 0,
    'AllRecvByteSize': 0,
    'ConnectionNum': 0,
    'ConnExecTimeSum': 0,
    'ConnExecTimeMin': 9999999999,
    'ConnExecTimeMax': 0,
}
conf = None


@gen.engine
def run():
    global testinfo, rtt_info, conf
    uri = testinfo['TargetURL']

    start = datetime.now()
    ws = yield websocket_connect(uri)
    end = datetime.now()

    # for RTT info
    if not rtt_info['TargetID']:
        rtt_info['TargetID'] = id(ws)

    connect_end_time = datetime.now()
    diff = end - start
    diff = diff.microseconds / 1000.    # msec
    testinfo['ConnectionNum'] += 1
    testinfo['ConnExecTimeSum'] += diff
    if testinfo['ConnExecTimeMin'] > diff:
        testinfo['ConnExecTimeMin'] = diff
    if testinfo['ConnExecTimeMax'] < diff:
        testinfo['ConnExecTimeMax'] = diff

    is_maxcon = False
    if testinfo['ConnectionNum'] >= conf['loops']:
        print("keep %d sec" % conf['keep'])
        is_maxcon = True

    def send_msg():
        packed_data = msgpack.packb([{'a': 1}, ])
        start = datetime.now()
        ws.write_message(packed_data, True)
        testinfo['AllSendByteSize'] += len(packed_data)
        if rtt_info['TargetID'] == id(ws):
            rtt_info['Start'] = start

    def wrap_send_msg():
        offset = random.randint(2, 7)
        IOLoop.instance().call_later(offset, send_msg)

    task = PeriodicCallback(wrap_send_msg, 3000)
    task.start()
    while True:
        msg = yield ws.read_message()
        end = datetime.now()
        testinfo['AllRecvByteSize'] += len(msg)
        if rtt_info['TargetID'] == id(ws):
            diff = end - rtt_info['Start']
            rtt_info['MsgNum'] += 1
            _tmp = diff.microseconds / 1000.
            rtt_info['Sum'] += _tmp
            if rtt_info['Min'] > _tmp:
                rtt_info['Min'] = _tmp
            if rtt_info['Max'] < _tmp:
                rtt_info['Max'] = _tmp
        diff = end - connect_end_time
        if diff.seconds > conf['keep'] and is_maxcon:
            break
    IOLoop.instance().stop()


def wrap_dump_info():
    global testinfo, rtt_info
    pid = os.getpid()

    def dump_info(pid):
        now = datetime.now()
        try:
            print("======= %s (elapsed: %s)" % (now, now - testinfo['StartTime']))
            print("tool's pid: %d" % pid)
            print("target url: %s" % testinfo['TargetURL'])
            print("Send Byte Size   : %s [byte] (%s)" % (
                humanize.intcomma(testinfo['AllSendByteSize']),
                humanize.naturalsize(testinfo['AllSendByteSize'])))
            print("Recive Byte Size : %s [byte] (%s)" % (
                humanize.intcomma(testinfo['AllRecvByteSize']),
                humanize.naturalsize(testinfo['AllRecvByteSize'])))
            print("Connection       : %d [conn]" % testinfo['ConnectionNum'])
            print("Connect Time(avg): %.1f [ms]" % (
                testinfo['ConnExecTimeSum']/testinfo['ConnectionNum']))
            print("Connect Time(min): %.1f [ms]" % testinfo['ConnExecTimeMin'])
            print("Connect Time(max): %.1f [ms]" % testinfo['ConnExecTimeMax'])
            print("Message RTT (avg): %.1f [ms]" % (
                rtt_info['Sum']/rtt_info['MsgNum']))
            print("Message RTT (min): %.1f [ms]" % rtt_info['Min'])
            print("Message RTT (max): %.1f [ms]" % rtt_info['Max'])
        except ZeroDivisionError:
            pass

    IOLoop.instance().call_later(0.1, dump_info, pid)


def main():
    global testinfo, conf
    parser = argparse.ArgumentParser(description='stress test tool for websocket servers')
    parser.add_argument('--config', dest='config_file', required=True, help='config file')
    optargs = parser.parse_args()
    conf = json.load(open(optargs.config_file))

    testinfo['StartTime'] = datetime.now()
    testinfo['TargetURL'] = conf['url'] + 'ws'
    for offset, cnt in enumerate(range(conf['loops'])):
        IOLoop.instance().call_later(offset * conf['interval'] / 1000., run)

    info_task = PeriodicCallback(wrap_dump_info, 10 * 1000)
    info_task.start()

    IOLoop.instance().start()


if __name__ == '__main__':
    main()
